src/components/ModalAuth.tsx

```diff
@@ -0,0 +1,203 @@
+    modalTitle: 'Sign Up'
+  },
+  resetPassword: {
+    name: 'resetPassword',
+    modalKey: 'resetPassword',
+    modalTitle: 'Reset Password'
+  }
+}
+
+/** Modal to leave feedback. */
+const ModalAuth = () => {
+
+  /** Checks if the given mode is actiev. */
+  const isModeActive = (mode: Mode) => activeMode.name === mode.name
+
+  const [activeMode, updateActiveMode] = useState(modes.signup)
+  const [email, updateEmail] = useState('')
+  const [password, updatePassword] = useState('')
+  const [error, updateError] = useState<null | string>(null)
+
+  const [isSubmitting, updateIsSubmitting] = useState(false)
+
+  const dispatch = useDispatch()
+
+  /** Sign up with email and password. */
+  const signUp: SubmitAction = useCallback(async (closeModal, email, password) => {
+
+    updateIsSubmitting(true)
+    try {
+      await window.firebase.auth().createUserWithEmailAndPassword(email, password!)
+      closeModal()
+      updateIsSubmitting(false)
+    }
+    catch (error) {
+      updateIsSubmitting(false)
+      return updateError(firebaseErrorsIndex[error?.code as errorCode] || firebaseErrorsIndex.default)
+    }
+  }, [])
+
+  const submitAction = useRef<SubmitAction>(signUp)
+
+  /** Handle email change. */
+  const onChangeEmail = (e: ChangeEvent<HTMLInputElement>) => {
+    updateEmail(e.target.value)
+  }
+
+  /** Handle password change. */
+  const onChangePassword = (e: ChangeEvent<HTMLInputElement>) => updatePassword(e.target.value)
+
+  /** Reset password using reset email. */
+  const resetPassword: SubmitAction = useCallback(async (closeModal, email) => {
+    updateIsSubmitting(true)
+
+    try {
+      await window.firebase.auth().sendPasswordResetEmail(email, { url: FIREBASE_REDIRECT_URL! })
+      updateIsSubmitting(false)
+    }
+    catch (error) {
+      updateIsSubmitting(false)
+      return updateError(error.message || firebaseErrorsIndex.default)
+    }
+    dispatch(alert('Please check your email'))
+    closeModal()
+  }, [])
+
+  /** Login with email and password. */
+  const loginWithEmailAndPassword: SubmitAction = useCallback(async (closeModal, email, password) => {
+    updateIsSubmitting(true)
+    try {
+      await window.firebase.auth().signInWithEmailAndPassword(email, password!)
+      updateIsSubmitting(false)
+      closeModal()
+    }
+    catch (error) {
+      updateIsSubmitting(false)
+      return updateError(firebaseErrorsIndex[error?.code as errorCode] || firebaseErrorsIndex.default)
+    }
+  }, [])
+
+  /**
+   * Reset Email and Password fields.
+   */
+  const resetFormFields = () => {
+    updateEmail('')
+    updatePassword('')
+    updateError(null)
+  }
+
+  /**
+   * Login using google account.
+   */
+  const signInWithGoogle = () => {
+    updateIsSubmitting(true)
+    dispatch(login())
+  }
+
+  useEffect(() => {
+    resetFormFields()
+    submitAction.current = isModeActive(modes.login) ? loginWithEmailAndPassword : isModeActive(modes.signup) ? signUp : resetPassword
```

Comment:
The state management is much better now with `modes`!

`submitAction` should also be refactored in a similar way. With the current architecture, it is treated as an independent variable when it is actually a dependent variable: it can always be derived from the active mode.

You can define `submitAction` as a small function that calls the appropriate function based on the active mode. Or you can perform the conditional logic directly in the render function rather than in `useEffect`.